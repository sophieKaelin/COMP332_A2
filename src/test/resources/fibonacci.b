//
// This file is part of COMP332 Assignment 2/3 2019.
//
// weBCPL, a retro BCPL to WebAssembly compiler.
//
// Â© 2019, Dominic Verity and Anthony Sloane, Macquarie University.
//         All rights reserved.
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.
//
// A simple iterative Fibonacci implementation in BCPL.
//

GLOBAL { 
    start:  1 
    writef: 94
}

LET Fibonacci(n) = VALOF 
{
    LET res1, res2, temp = 0, 1, ?

    WHILE 0 < n DO {
        temp := res1
        res1 := res2
        res2 := temp + res1
        n := n - 1
    }

  RESULTIS res1
}

LET start() BE 
{
    writef(
        "The 10th Fibonacci number is %n*n", 
        fibonacci(n))
}

