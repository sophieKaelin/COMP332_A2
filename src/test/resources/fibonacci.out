Program(
    Vector(
        GlobalDecl(
            Vector(
                GlobalEntry(IdnDef("start"), Some(IntExp(1))),
                GlobalEntry(IdnDef("writef"), Some(IntExp(94))))),
        LetDecl(
            Vector(
                LetFnClause(
                    IdnDef("Fibonacci"),
                    Vector(IdnDef("n")),
                    ValofExp(
                        Block(
                            Vector(
                                LetDecl(
                                    Vector(
                                        LetVarClause(
                                            Vector(IdnDef("res1"), IdnDef("res2"), IdnDef("temp")),
                                            Vector(IntExp(0), IntExp(1), UndefExp()))))),
                            Vector(
                                WhileDoStmt(
                                    LessExp(IntExp(0), IdnExp(IdnUse("n"))),
                                    Block(
                                        Vector(),
                                        Vector(
                                            AssignStmt(
                                                Vector(IdnExp(IdnUse("temp"))),
                                                Vector(IdnExp(IdnUse("res1")))),
                                            AssignStmt(
                                                Vector(IdnExp(IdnUse("res1"))),
                                                Vector(IdnExp(IdnUse("res2")))),
                                            AssignStmt(
                                                Vector(IdnExp(IdnUse("res2"))),
                                                Vector(
                                                    PlusExp(
                                                        IdnExp(IdnUse("temp")),
                                                        IdnExp(IdnUse("res1"))))),
                                            AssignStmt(
                                                Vector(IdnExp(IdnUse("n"))),
                                                Vector(MinusExp(IdnExp(IdnUse("n")), IntExp(1))))))),
                                ResultIsStmt(IdnExp(IdnUse("res1"))))))))),
        LetDecl(
            Vector(
                LetProcClause(
                    IdnDef("start"),
                    Vector(),
                    Block(
                        Vector(),
                        Vector(
                            CallStmt(
                                CallExp(
                                    IdnExp(IdnUse("writef")),
                                    Vector(
                                        StringExp(
                                            Vector(
                                                84,
                                                104,
                                                101,
                                                32,
                                                49,
                                                48,
                                                116,
                                                104,
                                                32,
                                                70,
                                                105,
                                                98,
                                                111,
                                                110,
                                                97,
                                                99,
                                                99,
                                                105,
                                                32,
                                                110,
                                                117,
                                                109,
                                                98,
                                                101,
                                                114,
                                                32,
                                                105,
                                                115,
                                                32,
                                                37,
                                                110,
                                                10)),
                                        CallExp(
                                            IdnExp(IdnUse("fibonacci")),
                                            Vector(IdnExp(IdnUse("n"))))))))))))))
